### PR TITLE
[settings] add some missing functionality required for add-on settings

### PR DESCRIPTION
**Pull request by Montellese**
## Description
Based on some feedback from @ronie quite a while ago I made some changes to the core's settings system to add some functionality supported by the old / outdated add-on settings system:
* support `<close>` property for action button `<control>`s
* support pure string based labels (non-translatable) for `<option>`
* extend supported `<masking>` format for path settings
* fix parsing of path setting with `<control type="button" format="image">`
* support `"sort"` attribute on `<options>` constraint of integer and string settings

## Motivation and Context
Make it easier for add-on authors to migrate from the old / outdated add-on settings system to the core's settings system.

## How Has This Been Tested?
I made some manual tests using the core's settings by changing some of the existing settings definitions. I was not able to test this with an actual add-on's settings. It would be great if someone could do that.

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by phunkyfish** [comment]
One I came across was when porting an addon to the new settings format was the use of float instead of integer. Not sure how common a use case that might be but could be a worthwhile addition to this PR.

**Comment by Montellese** [comment]
@phunkyfish could you be more specific? The core's settings system supports both integer and float (number) settings.

**Comment by phunkyfish** [comment]
Sure, let me dig out the example. I made a note of it somewhere.

**Comment by phunkyfish** [comment]
It may just have been that I tried to use float instead of number. Couldn’t find my note but will try again and report back.

**Comment by Montellese** [comment]
I just added another commit which should fix #15423. It was just a GUI / display problem. The setting's value was still saved and loaded properly.

**Comment by phunkyfish** [comment]
From old settings:

```
<setting id="epgTimeShift" type="slider" label="30024" default="0" range="-12,.5,14" option="float"/>
```

So I had to change to this:

```
        <setting id="epgTimeShift" type="integer" label="30024" help="30625">
          <level>0</level>
          <default>0</default>
          <constraints>
            <minimum>-720</minimum>
            <step>30</step>
            <maximum>840</maximum>
          </constraints>
          <control type="slider" format="integer">
            <formatlabel>14044</formatlabel>
          </control>
        </setting>
```

Tried both number and float. With float the setting does not appear and with number it crashes.

**Comment by phunkyfish** [comment]
Note that I could not find any mention of a `type=“number”` in the core’s settings.xml.

**Comment by Montellese** [comment]
You're right we don't use `type="number"` in core's `settings.xml` but it's still there see https://github.com/xbmc/xbmc/blob/master/xbmc/settings/lib/SettingsManager.cpp#L950 and https://github.com/xbmc/xbmc/blob/master/xbmc/settings/lib/Setting.h#L328-L369.

**Comment by ronie** [comment]
i'm getting a compile error:
https://paste.kodi.tv/yajubumuqe.kodi

**Comment by Montellese** [comment]
Argh stupid X11 defining `None`. I'll fix it ASAP.

**Comment by Montellese** [comment]
I've pushed a fix which should hopefully solve this for Linux and BSD. All other platforms are fine ;-)

File: xbmc/settings/lib/Setting.cpp [hunk 1]
```diff
@@ -42,6 +42,28 @@ bool CheckSettingOptionsValidity(const TValue& value, const std::vector<TKey>& o
   return false;
 }
 
+bool DeserializeOptionsSort(const TiXmlElement* optionsElement, SettingOptionsSort& optionsSort) {
```
Comment (Rechi) [inline xbmc/settings/lib/Setting.cpp]: > Curly braces always go on a new line.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#31-braces

File: xbmc/settings/lib/SettingDefinitions.h [hunk 2]
```diff
@@ -98,3 +98,9 @@ using StringSettingOptions = std::vector<StringSettingOption>;
 class CSetting;
 using IntegerSettingOptionsFiller = void (*)(std::shared_ptr<const CSetting> setting, IntegerSettingOptions &list, int &current, void *data);
 using StringSettingOptionsFiller = void (*)(std::shared_ptr<const CSetting> setting, StringSettingOptions &list, std::string &current, void *data);
+
+enum class SettingOptionsSort {
```
Comment (Rechi) [inline xbmc/settings/lib/SettingDefinitions.h]: > Curly braces always go on a new line.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#31-braces

File: xbmc/settings/windows/GUIControlSettings.cpp [hunk 3]
```diff
@@ -138,6 +150,21 @@ static bool GetIntegerOptions(SettingConstPtr setting, IntegerSettingOptions& op
     }
   }
 
+  switch (pSettingInt->GetOptionsSort())
+  {
+    case SettingOptionsSort::Ascending:
```
Comment (Rechi) [inline xbmc/settings/windows/GUIControlSettings.cpp]: > Do not indent `case` and `default`.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#332-switch-case

File: xbmc/settings/windows/GUIControlSettings.cpp [hunk 4]
```diff
@@ -138,6 +150,21 @@ static bool GetIntegerOptions(SettingConstPtr setting, IntegerSettingOptions& op
...
+      std::sort(options.begin(), options.end(), CompareSettingOptionAseconding<IntegerSettingOption>);
+      break;
+
+    case SettingOptionsSort::Descending:
```
Comment (Rechi) [inline xbmc/settings/windows/GUIControlSettings.cpp]: > Do not indent `case` and `default`.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#332-switch-case

File: xbmc/settings/windows/GUIControlSettings.cpp [hunk 5]
```diff
@@ -138,6 +150,21 @@ static bool GetIntegerOptions(SettingConstPtr setting, IntegerSettingOptions& op
...
+      std::sort(options.begin(), options.end(), CompareSettingOptionDeseconding<IntegerSettingOption>);
+      break;
+
+    case SettingOptionsSort::NoSorting:
```
Comment (Rechi) [inline xbmc/settings/windows/GUIControlSettings.cpp]: > Do not indent `case` and `default`.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#332-switch-case

File: xbmc/settings/windows/GUIControlSettings.cpp [hunk 6]
```diff
@@ -138,6 +150,21 @@ static bool GetIntegerOptions(SettingConstPtr setting, IntegerSettingOptions& op
...
+    default:
```
Comment (Rechi) [inline xbmc/settings/windows/GUIControlSettings.cpp]: > Do not indent `case` and `default`.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#332-switch-case

File: xbmc/settings/windows/GUIControlSettings.cpp [hunk 7]
```diff
@@ -196,6 +223,21 @@ static bool GetStringOptions(SettingConstPtr setting, StringSettingOptions& opti
       return false;
   }
 
+  switch (pSettingString->GetOptionsSort())
+  {
+    case SettingOptionsSort::Ascending:
```
Comment (Rechi) [inline xbmc/settings/windows/GUIControlSettings.cpp]: > Do not indent `case` and `default`.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#332-switch-case

File: xbmc/settings/windows/GUIControlSettings.cpp [hunk 8]
```diff
@@ -196,6 +223,21 @@ static bool GetStringOptions(SettingConstPtr setting, StringSettingOptions& opti
...
+      std::sort(options.begin(), options.end(), CompareSettingOptionAseconding<StringSettingOption>);
+      break;
+
+    case SettingOptionsSort::Descending:
```
Comment (Rechi) [inline xbmc/settings/windows/GUIControlSettings.cpp]: > Do not indent `case` and `default`.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#332-switch-case

File: xbmc/settings/windows/GUIControlSettings.cpp [hunk 9]
```diff
@@ -196,6 +223,21 @@ static bool GetStringOptions(SettingConstPtr setting, StringSettingOptions& opti
...
+      std::sort(options.begin(), options.end(), CompareSettingOptionDeseconding<StringSettingOption>);
+      break;
+
+    case SettingOptionsSort::NoSorting:
```
Comment (Rechi) [inline xbmc/settings/windows/GUIControlSettings.cpp]: > Do not indent `case` and `default`.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#332-switch-case

File: xbmc/settings/windows/GUIControlSettings.cpp [hunk 10]
```diff
@@ -196,6 +223,21 @@ static bool GetStringOptions(SettingConstPtr setting, StringSettingOptions& opti
...
+    default:
```
Comment (Rechi) [inline xbmc/settings/windows/GUIControlSettings.cpp]: > Do not indent `case` and `default`.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#332-switch-case

File: xbmc/addons/settings/AddonSettings.cpp [hunk 11]
```diff
@@ -1341,18 +1341,19 @@ bool CAddonSettings::ParseOldLabel(const TiXmlElement* element, const std::strin
...
-    return true;
-
-
-  // as a last resort use the setting's identifier as a label
-  bool parsed = !labelString.empty();
-  if (!parsed)
-    labelString = settingId;
+  if (parsed)
+  {
+    char *endptr;
```
Comment (Rechi) [inline xbmc/addons/settings/AddonSettings.cpp]: > Left-align `*` and `&` to the base type they modify.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#52-pointer-and-reference-types

**Comment by phunkyfish** [comment]
If I use number in the new format then this should work:

```
        <setting id="epgTimeShift" type="number" label="30024" help="30625">
          <level>0</level>
          <default>0</default>
          <constraints>
            <minimum>-12</minimum>
            <step>0.5</step>
            <maximum>14</maximum>
          </constraints>
          <control type="slider" format="number">
            <formatlabel>17998</formatlabel>
          </control>
        </setting>
```

However it crashes once the menu containing the Addon setting is accessed, here is the segfault and backtrace: https://pastebin.com/zybEGi20

File: xbmc/addons/settings/AddonSettings.cpp (hunk 11 above)
Comment (Montellese) [reply xbmc/addons/settings/AddonSettings.cpp]: This code has just been moved / indented from above. Should I still fix it? Personally I prefer to keep syntax changes out of functional changes.

File: xbmc/settings/windows/GUIControlSettings.cpp (hunk 3 above)
Comment (Montellese) [reply xbmc/settings/windows/GUIControlSettings.cpp]: I used the same style as present a few lines above but I can change it here.

File: xbmc/settings/lib/Setting.cpp (hunk 1 above)
Comment (Montellese) [reply xbmc/settings/lib/Setting.cpp]: Done

File: xbmc/settings/lib/SettingDefinitions.h (hunk 2 above)
Comment (Montellese) [reply xbmc/settings/lib/SettingDefinitions.h]: Done

File: xbmc/settings/windows/GUIControlSettings.cpp (hunk 3 above)
Comment (Montellese) [reply xbmc/settings/windows/GUIControlSettings.cpp]: Done

File: xbmc/settings/windows/GUIControlSettings.cpp (hunk 4 above)
Comment (Montellese) [reply xbmc/settings/windows/GUIControlSettings.cpp]: Done

File: xbmc/settings/windows/GUIControlSettings.cpp (hunk 5 above)
Comment (Montellese) [reply xbmc/settings/windows/GUIControlSettings.cpp]: Done

File: xbmc/settings/windows/GUIControlSettings.cpp (hunk 6 above)
Comment (Montellese) [reply xbmc/settings/windows/GUIControlSettings.cpp]: Done

File: xbmc/settings/windows/GUIControlSettings.cpp (hunk 7 above)
Comment (Montellese) [reply xbmc/settings/windows/GUIControlSettings.cpp]: Done

File: xbmc/settings/windows/GUIControlSettings.cpp (hunk 9 above)
Comment (Montellese) [reply xbmc/settings/windows/GUIControlSettings.cpp]: Done

File: xbmc/settings/windows/GUIControlSettings.cpp (hunk 10 above)
Comment (Montellese) [reply xbmc/settings/windows/GUIControlSettings.cpp]: Done

**Comment by Montellese** [comment]
@Rechi: Thanks for the review. I adjust (almost) all syntactic mistakes. Sorry, haven't been working in Kodi code for a while and we use a completely different style guide at work...

@phunkyfish: Looks like a bug somewhere in the GUI code (unrelated to add-on settings). I'll take a look. Thanks for the test setting.

**Comment by phunkyfish** [comment]
I found my notes. There was one other issue. When you have a setting of `type="path"` which has `<allowempty>true</allowempty` I could not figure out how to empty the value. After some messing about returning to the default folder does it but that wouldn't be obvious to a user. Same with controls of either `file` or `path`.

File: xbmc/addons/settings/AddonSettings.cpp (hunk 11 above)
Comment (Rechi) [reply xbmc/addons/settings/AddonSettings.cpp]: If you prefer to keep syntax changes in a separate commit add one before the actual code changes.
IMO it just spams the commit history and you could add it to the actual code change as the line is touched anyway.

Comment (Montellese) [reply xbmc/addons/settings/AddonSettings.cpp]: Fixed

**Comment by Montellese** [comment]
@phunkyfish could you please open new (separate) issues for the things you mentioned above? That should make it easier to track them. Thanks.

**Comment by phunkyfish** [comment]
Sure

**Comment by phunkyfish** [comment]
I created an issue for the real number not working problem: https://github.com/xbmc/xbmc/issues/16663

Turns out the second one is a non-issue as the delete key does clear the field as expected.

**Comment by Montellese** [comment]
I've added another commit that fixes action settings in the core's settings system for add-ons.

**Comment by Montellese** [comment]
I've added yet another commit to address #16663. We now catch the formatting exception thrown by `libfmt` and try formatting again with a fallback formatting string. I will take a look at the rest of the code because I'm pretty sure it might happen elsewhere too.

File: xbmc/settings/windows/GUIControlSettings.cpp [hunk 12]
```diff
@@ -36,6 +36,7 @@
 #include "settings/lib/Setting.h"
 #include "settings/lib/SettingDefinitions.h"
 #include "storage/MediaManager.h"
+#include "utils/log.h"
```
Comment (Rechi) [inline xbmc/settings/windows/GUIControlSettings.cpp]: move it two lines down

> Included header files have to be sorted (case sensitive) alphabetically [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#7-headers

Comment (Montellese) [reply xbmc/settings/windows/GUIControlSettings.cpp]: Fixed.

**Comment by ronie** [comment]
> * support `<close>` property for action button `<control>`s
> 
> * support pure string based labels (non-translatable) for `<option>`
> 
> * extend supported `<masking>` format for path settings
> 
> * fix parsing of path setting with `<control type="button" format="image">`
> 
> * support `"sort"` attribute on `<options>` constraint of integer and string settings

i've tested all of these and can confirm it fixes the issues i had before with those types of settings.
also the option to specify and action works ok now.

**Review by ronie — APPROVED** [review]

**Comment by Montellese** [comment]
From my POV this is ready. I will not add any more changes / fixes. I'm still contemplating how to (best) fix #15252. Once I've come up with a viable solution I'll open a new PR.

**Comment by Montellese** [comment]
How can I get this to move forward? Who is responsible for merges these days?

**Comment by phunkyfish** [comment]
I can merge this later today if someone else has not.

**Pull request merged**